README.markdown
# APND

APND (Apple Push Notification Daemon) is a ruby library to send Apple Push
Notifications to iPhones.

Apple recommends application developers create one connection to their
upstream push notification server, rather than creating one per notification.

APND acts as an intermediary between your application and Apple (see **APND
Daemon** below). Your application's notifications are queued to APND, which
are then sent to Apple over a single connection.

Within ruby applications, `APND::Notification` can be used to send
notifications to a running APND instance (see **APND Notification** below) or
directly to Apple. The command line can be used to send single notifications
for testing purposes (see **APND Client** below).


## General Usage

### APND Daemon

APND receives push notifications from your application and relays them to
Apple over a single connection as explained above. The `apnd` command line
utility is used to start APND.

    Usage:
      apnd daemon [OPTIONS] --apple-cert </path/to/cert>

    Required Arguments:
            --apple-cert      [PATH]     PATH to APN certificate from Apple

    Optional Arguments:
            --apple-host      [HOST]     Connect to Apple at HOST (default is gateway.sandbox.push.apple.com)
            --apple-port      [PORT]     Connect to Apple on PORT (default is 2195)
            --apple-cert-pass [PASSWORD] PASSWORD for APN certificate from Apple
            --daemon-port     [PORT]     Run APND on PORT (default is 22195)
            --daemon-bind     [ADDRESS]  Bind APND to ADDRESS (default is 0.0.0.0)
            --daemon-log-file [PATH]     PATH to APND log file (default is /var/log/apnd.log)
            --daemon-timer    [SECONDS]  Set APND queue refresh time to SECONDS (default is 30)
            --foreground                 Run APND in foreground without daemonizing

    Help:
            --help                       Show this message


### APND Client

APND includes a command line client which can be used to send notifications to
a running APND instance. It is only recommended to send notifications via
`apnd push` for testing purposes.

    Usage:
      apnd push [OPTIONS] --token <token>

    Required Arguments:
            --token  [TOKEN]             Set Notification's iPhone token to TOKEN

    Optional Arguments:
            --alert  [MESSAGE]           Set Notification's alert to MESSAGE
            --sound  [SOUND]             Set Notification's sound to SOUND
            --badge  [NUMBER]            Set Notification's badge number to NUMBER
            --custom [JSON]              Set Notification's custom data to JSON
            --host   [HOST]              Send Notification to HOST, usually the one running APND (default is 'localhost')
            --port   [PORT]              Send Notification on PORT (default is 22195)

    Help:
            --help                       Show this message


### APND Notification

The `APND::Notification` class can be used within your application to send
push notifications to APND.

    require 'apnd'

    # Set the host/port APND is running on
    # (not needed if you're using localhost:22195)
    # Put this in config/initializers/apnd.rb for Rails
    APND::Notification.upstream_host = 'localhost'
    APND::Notification.upstream_port = 22195


    # Initialize some notifications
    notification1 = APND::Notification.new(
      :alert  => 'Alert!',
      :token  => 'fe15a27d5df3c34778defb1f4f3880265cc52c0c047682223be59fb68500a9a2',
      :badge  => 1
    )

    notification2 = APND::Notification.new(
      :alert  => 'Red Alert!',
      :token  => 'fe15a27d5df3c34778defb1f4f3880265cc52c0c047682223be59fb68500a9a2',
      :badge  => 99
    )


    # Send multiple notifications at once to avoid overhead in
    # opening/closing the upstream socket connection each time
    #
    # *IMPORTANT!* Use sock.puts as it appends a new line. If you don't,
    # you'll only receive the first notification.

    APND::Notification.open_upstream_socket do |sock|
      sock.puts(notification1.to_bytes)
      sock.puts(notification2.to_bytes)
    end


    # Send a notification to the upstream socket immediately
    notification3 = APND::Notification.create(
      :alert  => 'Alert!',
      :token  => 'fe15a27d5df3c34778defb1f4f3880265cc52c0c047682223be59fb68500a9a2',
      :badge  => 0
    )


### APND Feedback

Apple Push Notification Service keeps a log when you attempt to deliver
a notification to a device that has removed your application. A Feedback
Service is provided which applications should periodically check to remove
from their databases.

The `APND::Feedback` class can be used within your application to retrieve
a list of device tokens that you are sending notifications to but have
removed your application.

    APND::Feedback.upstream_host = 'feedback.push.apple.com'
    APND::Feedback.upstream_port = 2196

    # Block form
    APND::Feedback.find_stale_devices do |token, removed_at|
      device = YourApp::Device.find_by_token(token)
      unless device.registered_at > removed_at
        device.push_enabled = 0
        device.save
      end
    end

    # Array form
    stale = APND::Feedback.find_stale_devices
    stale.each do |(token, removed_at)|
      device = YourApp::Device.find_by_token(token)
      unless device.registered_at > removed_at
        device.push_enabled = 0
        device.save
      end
    end


## Prerequisites

You must have a valid Apple Push Notification Certificate for your iPhone
application. Obtain your APN certificate from the iPhone Provisioning Portal
at [developer.apple.com](http://developer.apple.com/).


## Requirements

* [EventMachine](http://github.com/eventmachine/eventmachine)
* [Daemons](http://github.com/ghazel/daemons)
* [JSON](http://github.com/flori/json)

Ruby must be compiled with OpenSSL support.

Tests use [Shoulda](http://github.com/thoughtbot/shoulda), and optionally
[TURN](https://github.com/TwP/turn).


## Installation

RubyGems:

    gem install apnd

Git:

    git clone git://github.com/itspriddle/apnd.git


## Credit

APND is based on [apnserver](http://github.com/bpoweski/apnserver) and
[apn_on_rails](http://github.com/PRX/apn_on_rails). Either worked just how I
wanted, so I rolled my own using theirs as starting points. If APND doesn't
suit you, check them out instead.


## Contributors

* Mark Wotton ([mwotton](https://github.com/mwotton))


## Copyright

Copyright (c) 2010-2011 Joshua Priddle. See LICENSE for details.


HISTORY.markdown
# 0.2.0 - 2011-04-04

Note that v0.2.0 fixes a critical error, and that previous releases of APND
should not be used.

* Fixes a critical error in `APND::Daemon::Protocol` which caused problems
  with notifications containing new line characters in their message. Props to
  [mwotton](https://github.com/mwotton).
* Merged `apnd` and `apnd-push` CLI tools into `apnd`.
* Added support for Apple Push Notification Feedback Service


lib/apnd/cli.rb
require 'daemons'
require 'optparse'

module APND
  class CLI #:nodoc: all

    #
    # Run apnd push
    #
    def self.push(argv)
      help = <<-HELP
Usage:
  apnd push [OPTIONS] --token <token>

      HELP

      options = {}

      opts = OptionParser.new do |opt|
        opt.banner = help

        opt.separator "Required Arguments:\n"

        opt.on('--token  [TOKEN]', "Set Notification's iPhone token to TOKEN") do |token|
          options[:token] = token
        end

        opt.separator "\nOptional Arguments:\n"

        opt.on('--alert  [MESSAGE]', "Set Notification's alert to MESSAGE") do |alert|
          options[:alert] = alert
        end

        opt.on('--sound  [SOUND]', "Set Notification's sound to SOUND") do |sound|
          options[:sound] = sound
        end

        opt.on('--badge  [NUMBER]', "Set Notification's badge number to NUMBER") do |badge|
          options[:badge] = badge.to_i
        end

        opt.on('--custom [JSON]', "Set Notification's custom data to JSON") do |custom|
          begin
            options[:custom] = JSON.parse(custom)
          rescue JSON::ParserError => e
            puts "Invalid JSON: #{e}"
            exit -1
          end
        end

        opt.on('--host   [HOST]', "Send Notification to HOST, usually the one running APND (default is 'localhost')") do |host|
          options[:host] = host
        end

        opt.on('--port   [PORT]', 'Send Notification on PORT (default is 22195)') do |port|
          options[:port] = port.to_i
        end

        opt.separator "\nHelp:\n"

        opt.on('--help', 'Show this message') do
          puts opt
          exit
        end
      end

      begin
        opts.parse!
        if options.empty?
          puts opts
          exit
        end

        unless options[:token]
          raise OptionParser::MissingArgument, "must specify --token"
        end
      rescue OptionParser::InvalidOption, OptionParser::MissingArgument
        puts "#{$0}: #{$!.message}"
        puts "#{$0}: try '#{$0} --help' for more information"
        exit
      end

      # Configure Notification upstream host/port
      APND::Notification.upstream_host = options.delete(:host) if options[:host]
      APND::Notification.upstream_port = options.delete(:port) if options[:port]

      APND::Notification.create(options)
    end

    #
    # Run apnd daemon
    #
    def self.daemon(argv)
      help = <<-HELP
Usage:
  apnd daemon --apple-cert </path/to/cert>

      HELP

      options = {}

      opts = OptionParser.new do |opt|
        opt.banner = help

        opt.separator "Required Arguments:\n"

        opt.on('--apple-cert      [PATH]', 'PATH to APN certificate from Apple') do |cert|
          options[:apple_cert] = cert
        end

        opt.separator "\nOptional Arguments:\n"

        opt.on('--apple-host      [HOST]', "Connect to Apple at HOST (default is gateway.sandbox.push.apple.com)") do |host|
          options[:apple_host] = host
        end

        opt.on('--apple-port      [PORT]', 'Connect to Apple on PORT (default is 2195)') do |port|
          options[:apple_port] = port.to_i
        end

        opt.on('--apple-cert-pass [PASSWORD]', 'PASSWORD for APN certificate from Apple') do |pass|
          options[:apple_cert_pass] = pass
        end

        opt.on('--daemon-port     [PORT]', 'Run APND on PORT (default is 22195)') do |port|
          options[:daemon_port] = port.to_i
        end

        opt.on('--daemon-bind     [ADDRESS]', 'Bind APND to ADDRESS (default is 0.0.0.0)') do |bind|
          options[:daemon_bind] = bind
        end

        opt.on('--daemon-log-file [PATH]', 'PATH to APND log file (default is /var/log/apnd.log)') do |log|
          options[:daemon_log_file] = log
        end

        opt.on('--daemon-timer    [SECONDS]', 'Set APND queue refresh time to SECONDS (default is 30)') do |seconds|
          options[:daemon_timer] = seconds.to_i
        end

        opt.on('--foreground', 'Run APND in foreground without daemonizing') do
          options[:foreground] = true
        end

        opt.separator "\nHelp:\n"

        opt.on('--help', 'Show this message') do
          puts opt
          exit
        end
      end

      begin
        opts.parse!
        if options.empty?
          puts opts
          exit
        end

        unless options[:apple_cert]
          raise OptionParser::MissingArgument, "must specify --apple-cert"
        end
      rescue OptionParser::InvalidOption, OptionParser::MissingArgument
        puts "#{$0}: #{$!.message}"
        puts "#{$0}: try '#{$0} --help' for more information"
        exit
      end

      APND.configure do |config|
        # Setup AppleConnection
        config.apple.cert      = options[:apple_cert]      if options[:apple_cert]
        config.apple.cert_pass = options[:apple_cert_pass] if options[:apple_cert_pass]
        config.apple.host      = options[:apple_host]      if options[:apple_host]
        config.apple.port      = options[:apple_port]      if options[:apple_port]

        # Setup Daemon
        config.daemon.bind     = options[:daemon_bind]     if options[:daemon_bind]
        config.daemon.port     = options[:daemon_port]     if options[:daemon_port]
        config.daemon.log_file = options[:daemon_log_file] if options[:daemon_log_file]
        config.daemon.timer    = options[:daemon_timer]    if options[:daemon_timer]
      end

      if APND.settings.apple.cert.nil?
        puts opts
        exit
      else
        unless options[:foreground]
          Daemonize.daemonize(APND.settings.daemon.log_file, 'apnd')
        end
        APND::Daemon.run!
      end

    end
  end
end
